File: 2018/07-jul/7-29-overloading/javascript/main.js
const Animal = require("./Animal");
const Plant = require("./Plant");
const Garden = require("./Garden");

/**
 * Create a garden to demonstrate method overloading in JavaScript
 * @author Andrew Jarombek
 * @since 7/27/2018
 */

/**
 * Test a certain case for JavaScript truth.  Assertions are a nice form of inline documentation
 * @param assertion - a statement that will be checked whether or not it is truthy
 */
const assert = (assertion) => {
    console.assert(assertion, `Assertion failed!`);
};

const msGroundhog = new Animal("Ms. Groundhog", Animal.Species.GROUNDHOG,
    "Enjoys lounging and eating grass all day");
const mrGroundhog = new Animal("Mr. Groundhog", Animal.Species.GROUNDHOG);

const doe = new Animal("doe", Animal.Species.DEER,
    "from the tip of his wand burst the silver doe");
const bunny = new Animal("bunny", Animal.Species.RABBIT);

const hosta = new Plant(Plant.Species.HOSTA);
const lily = new Plant(Plant.Species.DAYLILY, true);
const iris = new Plant(Plant.Species.IRIS, false);

const garden = new Garden([msGroundhog, mrGroundhog, doe, bunny], [hosta, lily, iris]);

// Both checks on animals and plants will succeed
assert(garden.inGarden(doe));
assert(garden.inGarden(lily));

assert(!garden.inGarden(new Animal("momma rabbit", Animal.Species.RABBIT)));

File: 2018/07-jul/7-29-overloading/python/main.py
#!/usr/bin/env python

"""
Author: Andrew Jarombek
Date: 7/28/2018
"""

from Animal import Animal, AnimalSpecies
from Plant import Plant, PlantSpecies
from Garden import Garden

msGroundhog = Animal("Ms. Groundhog", AnimalSpecies.GROUNDHOG,
                     description="Enjoys lounging and eating grass all day")
mrGroundhog = Animal("Mr. Groundhog", AnimalSpecies.GROUNDHOG)

doe = Animal("doe", AnimalSpecies.DEER,
             description="from the tip of his wand burst the silver doe")
bunny = Animal("bunny", AnimalSpecies.RABBIT)

hosta = Plant(PlantSpecies.HOSTA)
lily = Plant(PlantSpecies.DAYLILY, in_bloom=True)
iris = Plant(PlantSpecies.IRIS, in_bloom=False)

garden = Garden(animals=[msGroundhog, mrGroundhog, doe, bunny],
                plants=[hosta, lily, iris])

assert garden.in_garden(doe)
assert garden.in_garden(lily)
assert not garden.in_garden(Animal("momma rabbit", AnimalSpecies.RABBIT))
